Put comma after any non-whitespace non-comment characters in XJSExpression

jasom opened the pull request:
For some reason the comma was emitted in renderXJSExpressionContainer if
it was an XJSExpressionContainer, and in the caller of the function
otherwise.  This made getting things like `<Foo bar=(baz) quux={biff} />`
hard.

This should fix issue #1673

jasom commented:
Re: the TravisCI failure:

Hmm; that test passed locally; I'll check if I'm missing anything from the commit, and update.

jasom commented:
Added in the missing change, and signed cla.

zpao commented:
Hmm, can you run `npm test` locally - it looks like that might not be working on Travis (the `grunt jest` task doesn't seem to be outputting anything and is passing but running locally it fails). This will run our transform tests (we should also add one for whatever changes we have here, looks like there was a test case in the linked issue).

jasom commented:
Ah, `npm test` is much better; I was doing a `grunt test` previously to test it, per the readme.  It looks like it wants the comma before the following whitespace.  Is this a correctness issue with ASI in the event of a newline, or just an aesthetic thing?

jasom commented:
The tests seem to think that for:
```
<foo> {
 bar
//comment
}
{ baz}
</foo>
```
The comma ought go immediately after bar.  That surprises me; does anyone happen to know why this is the case?  I can sort-of see putting the comma after the last non-whitespace item in the braces, but putting it before comments seems rather odd to me.

syranide commented:
The reason for them being inline into the sub visitors is to have control over where they end up. Moving it out like that works, but you then end up with commas after newlines/whitespace rather than before as you would write yourself.

jasom commented:
@syranide Starting to get a bit off topic, but as far as "as you would write yourself." you probably wouldn't put a space right before a newline, so that becomes a question of how far you want to take it.  I'm testing a patch that fixes #1673 without changing any existing output; it's a lot more complicated than this one though.

syranide commented:
@jasom I don't see how this is off-topic, the current output by JSX (except for whitespace before newline) was agreed on in #970 and follows common style-guides/lints and the way you would write it yourself, this PR seems to undo just that?

`, \n` is a deficiency in the current implementation due to the complexity of solving it (it seems).

jasom commented:
@syranide I meant the `, \n` nitpick was off topic, not your comment.

syranide commented:
@jasom Ah :)

jasom commented:
Okay, I fixed the same issue in a completely different way; it should change not at all any whitespace placement.  I also added a unit test for the issue I fixed.

syranide commented:
@jasom Your commit is broken (it includes git comments).

syranide commented on vendor/fbtransform/transforms/xjs.js:
```diff
@@ -11,6 +11,61 @@
+    var commaPos = 0;
+    //console.log('"'+value+'"')
+    for (var i=0;i<value.length;++i) {
+        switch (state) {
```
My vote goes to `if` instead of this nested `switch` nightmare.

jasom commented:
@syranide Sorry, I pushed the wrong branch

jasom commented:
@syranide Fixed up git garbage, converted switch to if/else.  I have to go to work now so any other changes will have to wait until tonight.

zpao commented:
cc @jeffmo

syranide commented on vendor/fbtransform/transforms/xjs.js:
```diff
@@ -11,6 +11,51 @@
+    var state = "normal";
+    var commaPos = 0;
+    for (var i=0;i<value.length;++i) {
+        if (state === "normal") {
```
You have 3 else-branches with the same `commaPos = i+1;`, it should be possible to reduce and flatten this unless there's a very good reason for this verbosity.

jasom replied:
Yeah, the state machine started out a bit differently, I'll clean that up.

zpao replied:
Couple small style nits (can fixup later with anything else that might come):
- 2 space indent (match the rest of the files)
- spaces around operators (`for (var i = 0; i < value.length; ++i)`)
- use `===` and semicolons consistently. The line below this is missing the former. Line 28 the latter.

jasom replied:
@zpao I fixed those in all the places I found; is there a style linter I can use to check for this in the future?

hzoo replied:
You can run eslint with `grunt eslint` or `grunt lint`

jasom replied:
@hzoo Yeah, I did that, but it didn't flag any of the style issues zpao pointed out.

hzoo replied:
Oh then it's probably because it's not being processed - yeah `src/vendor` is in the `eslintignore`.

```
# We can probably lint these later but not important at this point
src/vendor
```

 You could install it globally and run it on a single file or remove the ignore temporarily.

jasom commented:
@zpao I fixed that in all the places I saw; is there a style linter I can use in the future to check for those?

jasom commented:
Is there anything else I need do to get this PR ready for merge?

zpao commented:
There are still a few stylistic things. I'm in the process of enabling eslint here so that these files get linted and can tell you what's going on (namely use single quotes, match other styles with indentation, else bracing). If you rebase on top of #3206 (or it will be in master shortly I think) then you can see this yourself with `grunt lint` (that grunt task doesn't take params and is actually a bit noisy about long lines in other code so you could also run `./node_modules/.bin/eslint vendor/fbtransform` directly)

jasom commented:
The changes I made should be clean to eslint now (I ran it directly on
the file last night).  Do you want me to clean up the entire jsx.js?

-Jason
On 15:34 Thu 19 Feb     , Paul O’Shannessy wrote:
>    There are still a few stylistic things. I'm in the process of enabling
>    eslint here so that these files get linted and can tell you what's
>    going on (namely use single quotes, match other styles with
>    indentation, else bracing). If you rebase on top of [1]#3206 (or it
>    will be in master shortly I think) then you can see this yourself with
>    grunt lint (that grunt task doesn't take params and is actually a bit
>    noisy about long lines in other code so you could also run
>    ./node_modules/.bin/eslint vendor/fbtransform directly)
> 
>    —
>    Reply to this email directly or [2]view it on GitHub.
> 
> References
> 
>    1. https://github.com/facebook/react/pull/3206
>    2. https://github.com/facebook/react/pull/3129#issuecomment-75162000

zpao commented:
Please rebase and re-run. If you were running eslint from your branch point, it wasn't configured correctly. It has since been configured appropriately and I cleaned up everything else in the above PR. The remaining errors and warnings are a result of your changes (image below is with your branch rebased on master).
![screen shot 2015-02-19 at 6 44 29 pm](https://cloud.githubusercontent.com/assets/8445/6280731/7080d27c-b867-11e4-9a3f-86630213dded.png)

zpao commented on vendor/fbtransform/transforms/xjs.js:
```diff
@@ -84,11 +121,13 @@ function renderXJSExpressionContainer(traverse, object, isLast, path, state) {
     utils.catchup(object.expression.range[1], state, trimLeft);
-    utils.append(', ', state);
+    utils.catchup(object.range[1] -1, state, commaAfterLastParen)
+    //utils.append(', ', state);
```
Lint doesn't catch it but please remove this.

jasom replied:
How embarrassing!  I removed it now.

jasom commented:
Thanks @zpao after rebase fixing the lint issues was easy.

jasom commented:
Let me know if you want me to squash some of these cleanup commits together

syranide commented on vendor/fbtransform/transforms/xjs.js:
```diff
@@ -11,6 +11,42 @@
+  var commaPos = 0;
+  for (var i = 0; i < value.length; ++i) {
+    if (state === 'normal') {
+      if (value.charAt(i) === '/') {
```
State-machine type of deal and all that... but this could be reduced to `value.substr(i, 2) === '//'`, etc. @zpao?

jeffmo replied:
+1

jasom replied:
I suppose it wouldn't surprise you that my day job uses mostly C...

jasom commented:
Any other changes needed?

jasom commented:
Just checking in on he status of this.

jasom commented:
Just checking in to see if any action is needed by me on this.

zpao commented:
Sorry, been focused on shipping 0.13. This won't make that cut but I'll come back to it after. In the mean time, this doesn't merge cleanly (probably mostly just because xjs.js is now jsx.js) so getting it back into a mergable state will help.

jasom commented:
Now rebased on top of master

jasom commented:
0.13 is out now.  @zpao: Let me know if I'm being too annoying.

zpao commented:
Squashed and landed as ef796790ec80c200bfc99df69e23db2522479272. Thanks and I'm sorry it took as long as it did!
